### PR TITLE
Add NotFound handler

### LINKED ISSUE #18060: Trailing slash leads to 404 (regression?)
### Gitea Version

3a77465e4ed252f5d92da12321613328b30b941c

### Git Version

2.30.2 (not relevant?)

### Operating System

Debian GNU/Linux 11

### How are you running Gitea?

local development via make

### Database

SQLite

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Description

Release 1.15.8 might contain a regression. At Codeberg.org we were made aware by users that a trailing slash in the URL leads to a simple "404 page not found" now, breaking some links on the site. If this route is no longer to be used, a forward would be a good option instead.

I made a bisect:

~~~
3a77465e4ed252f5d92da12321613328b30b941c is the first bad commit
commit 3a77465e4ed252f5d92da12321613328b30b941c
Author: zeripath <art27@cantab.net>
Date:   Thu Dec 16 23:03:20 2021 +0000

    Prevent double decoding of % in url params  (#17997) (#18001)
~~~
So 3a77465e4ed252f5d92da12321613328b30b941c (#17997) (#18001) are relevant.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by zeripath**
PR #17997 means that urls with terminal '/' are no longer immediately mapped
to the url without a terminal slash. However, it has revealed that the NotFound handler
appears to have been lost.

This PR adds back in a NotFound handler that simply redirects to a path without the
terminal slash or runs the NotFound handler.

Fix #18060

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: routers/web/web.go [hunk 1]
```diff
@@ -1078,4 +1078,13 @@ func RegisterRoutes(m *web.Route) {
 	if setting.API.EnableSwagger {
 		m.Get("/swagger.v1.json", SwaggerV1Json)
 	}
+	m.NotFound(func(w http.ResponseWriter, req *http.Request) {
+		escapedPath := req.URL.EscapedPath()
+		if len(escapedPath) > 1 && escapedPath[len(escapedPath)-1] == '/' {
+			http.Redirect(w, req, setting.AppSubURL+escapedPath[:len(escapedPath)-1], http.StatusFound)
```
Comment (KN4CK3R) [inline routers/web/web.go]:
```suggestion
			http.Redirect(w, req, setting.AppSubURL+escapedPath[:len(escapedPath)-1], http.StatusTemporaryRedirect)
```

Comment (zeripath) [reply routers/web/web.go]: All of our other Redirects are StatusFound

Comment (wxiaoguang) [reply routers/web/web.go]: Would using `escapedPath` cause problems for URLs like `/a%20b` ?

Comment (lunny) [reply routers/web/web.go]: We may need more testcases.

Comment (lunny) [reply routers/web/web.go]: And we need a return here.

Comment (wxiaoguang) [reply routers/web/web.go]: We forgot the `return` too many times. I would propose to make the framework like this:

```
func Handler(...) (Responser) {
    return ctx.NotFound(...)
}
```

Comment (zeripath) [reply routers/web/web.go]: > Would using `escapedPath` cause problems for URLs like `/a%20b` ?

Why would escape path cause problems? It's precisely what you should use in that case!

File: routers/web/web.go [hunk 2]
```diff
@@ -1078,4 +1078,13 @@ func RegisterRoutes(m *web.Route) {
...
+		}
```
Comment (zeripath) [inline routers/web/web.go:1086]:
```suggestion
				return
		}
````

File: routers/web/web.go (hunk 1 above)
Comment (zeripath) [reply routers/web/web.go]: > 

Return has been added.

Comment (wxiaoguang) [reply routers/web/web.go]: > > Would using `escapedPath` cause problems for URLs like `/a%20b` ?
> 
> Why would escape path cause problems? It's precisely what you should use in that case!

~~No, I do not think it is the correct case. The URL in `Location` header should be encoded correctly.~~

Comment (wxiaoguang) [reply routers/web/web.go]: Oh, you are right. The escaped path doesn't decode other encodings.

**Review by wxiaoguang — APPROVED** [review]

File: routers/web/web.go (hunk 1 above)
Comment (KN4CK3R) [reply routers/web/web.go]: > All of our other Redirects are StatusFound

That does not make this one right.

MDN:

> Even if the specification requires the method (and the body) not to be altered when the redirection is performed, not all user-agents conform here - you can still find this type of bugged software out there. It is therefore recommended to set the 302 code only as a response for GET or HEAD methods and to use 307 Temporary Redirect instead, as the method change is explicitly prohibited in that case.

> The only difference between 307 and 302 is that 307 guarantees that the method and the body will not be changed when the redirected request is made. With 302, some old clients were incorrectly changing the method to GET: the behavior with non-GET methods and 302 is then unpredictable on the Web, whereas the behavior with 307 is predictable. For GET requests, their behavior is identical. 

A POST to `/foo/` may become a GET `/foo` with the current code.

**Review by lunny — APPROVED** [review]

File: routers/web/web.go [hunk 3]
```diff
@@ -1078,4 +1078,14 @@ func RegisterRoutes(m *web.Route) {
 	if setting.API.EnableSwagger {
 		m.Get("/swagger.v1.json", SwaggerV1Json)
 	}
+	m.NotFound(func(w http.ResponseWriter, req *http.Request) {
+		escapedPath := req.URL.EscapedPath()
+		if len(escapedPath) > 1 && escapedPath[len(escapedPath)-1] == '/' {
+			http.Redirect(w, req, setting.AppSubURL+escapedPath[:len(escapedPath)-1], http.StatusFound)
```
Comment (zeripath) [inline routers/web/web.go]:
```suggestion
			http.Redirect(w, req, setting.AppSubURL+escapedPath[:len(escapedPath)-1], http.StatusTemporaryRedirect)
```

File: routers/web/web.go (hunk 1 above)
Comment (zeripath) [reply routers/web/web.go]: Good catch @KN4CK3R

Comment (zeripath) [reply routers/web/web.go]: PR updated

**Comment by zeripath** [comment]
make lgtm work

**Pull request merged**